### PR TITLE
Add Synchronous Clocks FMUs and example

### PR DESCRIPTION
**Pull request by t-sommer**

### CODE REVIEW

**Comment by clagms** [comment]
The example, kindly ported by @t-sommer from https://github.com/modelica/Reference-FMUs/pull/193,  consists of a controller and a plant, and a supervisor FMU. It uses model exchange in FMI3.0. The Controller fmu declares an input periodic clock, and the supervisor has an output clock that triggers when a state event occurs. The output clock of the supervisor is connected to another input clock of the controller.

The key part of the example is to show that both clocks can activate at the same time, if the conditions are right (and the conditions are right, at least on my machine). 
When that happens, the dependencies between the variables must be taken into account, in order to correctly propagate the values for the clocked partitions.

The supervisor monitors the plant’s continuous output and when a particular event happens, activates a clock and changes one of the controller’s inputs, which in turn change the controller algorithm.
The plant is just solving a basic ODE with an input from the controller. 

Example plot produced:
image

(blue – controller output)
(orange – plant output)
The oscillatory behavior is caused by the supervisor’s state events, as it changes the direction of u’s growth.

**Comment by clagms** [comment]
@t-sommer I think this PR is ready for review, except that the linter is complaining of my asciiart (at the head of `synchronous_control_me.c`).

@andreas-junghanns rightly asked for some documentation on the example, that's why the drawing is there.

I could move it to a readme file, but it's unclear where that file should reside.

Is there a way to tell the linter to ignore a portion of a file (or, even better, ignore what's commented out)?

**Comment by clagms** [comment]
Aside from the linter issues, I think it's time to figure out what to do with this example.
@t-sommer have you reconsidered whether this should be merged into the reference FMUs? I completely understand the argument of maintainability.
There's no doubt that there is a need for these examples, so we just need to figure out where they should reside.

**Comment by andreas-junghanns** [comment]
@clagms : Sorry, but I think @t-sommer is better to review here.

File: Supervisor/model.c [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        default:
+            logError(comp, "Unexpected value reference: %d.", vr);
+            return Error;
+    }
+}
+
+Status getClock(ModelInstance* comp, ValueReference vr, bool* value) {
+    switch (vr) {
+        case vr_s:
+            if (comp->state == EventMode && M(stateEvent)) {
```
Comment (t-sommer) [inline Supervisor/model.c:47]: With this code `value` is always `true` for `vr_s`.

File: Supervisor/model.c [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    }
+}
+
+void enterEventMode(ModelInstance* comp, int stepEvent, int stateEvent, const int rootsFound[], size_t nEventIndicators, int timeEvent) {
+    UNUSED(stepEvent);
+    UNUSED(rootsFound);
+    UNUSED(nEventIndicators);
+    UNUSED(timeEvent);
+
+    M(stateEvent) = stateEvent == 1;
```
Comment (t-sommer) [inline Supervisor/model.c:66]: This is not robust if `stateEvent == 2` (fmi3EventUnknown).

Comment (masoud-najafi) [reply Supervisor/model.c:66]: Hi @t-sommer @clagms 
I cannot (or I do not know) how to retrieve the FMUs from here. Would it be possible to zip them and share it here?
Then I could test them on my side.
Thanks,
Masoud

Comment (t-sommer) [reply Supervisor/model.c:66]: You can just [download the this branch](https://github.com/t-sommer/Reference-FMUs/archive/refs/heads/synchronous-clocks.zip) and follow the instructions in the [readme](https://github.com/modelica/Reference-FMUs#build-the-fmus).

**Comment by clagms** [comment]
TODO: Address comment made by @KarlWernersson here: https://github.com/modelica/fmi-standard/issues/1739#issuecomment-1091742698

**Comment by masoud-najafi** [comment]
Hi,
I checked the FMU Controller.fmu. I get this error.
 fmi3GetIntervalDecimal has not been implemented for reference=1.
Has this been fixed?

**Comment by clagms** [comment]
Hi Masoud! I'm fixing that problem, as well as other problems caused by the changes in the main branch. Will come back to you when it's fixed.

**Comment by clagms** [comment]
PS - @masoud-najafi  pls let me know if you have any trouble making this scenario work on your end. Simon (@SimplisticCode) was able to load them into his tool with no problems (other than the ones you pointed out).

**Comment by clagms** [comment]
@t-sommer is there a way to add an exception to the linting process? I need the non-ascii characters for the drawing of the scenario at the top of the file.

**Comment by masoud-najafi** [comment]
Thanks @clagms,  I need to move to 3.0 from Beta.4. I will check then. Hope it will not take much time.
If you specify the exact scenario, that produces the above result, it would be very good.

**Comment by clagms** [comment]
The scenario is drawn and documented in https://github.com/t-sommer/Reference-FMUs/blob/synchronous-clocks/examples/synchronous_control_me.c

After compiling the project, run the binary `synchronous_control_me` from its current directory and it should generate a csv called  `synchronous_control_me_out.csv`.
Create a basic plot in excel with those results and it should look like the above plot.

**Comment by t-sommer** [comment]
> @t-sommer is there a way to add an exception to the linting process? I need the non-ascii characters for the drawing of the scenario at the top of the file.

This should be converted to an SVG and added to the documentation, so users can see it when reading the documentation.

**Comment by clagms** [comment]
@KarlWernersson , @masoud-najafi , and @MBlesken , I think this example is ready for review.

Ideally, it would be nice if you could reproduce the results in your importer, using the Plant, Controller, and Supervisor fmus. I think @masoud-najafi is attempting this. 

Thank you!

**Comment by masoud-najafi** [comment]
HI @clagms ,
I get this error: fmi3GetShiftDecimal is not defined. I will try to fix it on my side and testing again.

**Comment by clagms** [comment]
Thanks @masoud-najafi . Let me know what the solution is. Probably we just need to define the function and output zero.

**Comment by masoud-najafi** [comment]
Yes, that's what I did.
Now, in the superviser.fmu you need to define GetFloat64 for reference vr_x (2).
I keep checking.

**Comment by masoud-najafi** [comment]
Also in plant.fmu,: GetFloat64 should accept u_r.

**Comment by masoud-najafi** [comment]
I know, x is an input and it does not make sense to retrieve it. Maybe I am wrong, but do you see somewhere in the standard that it is not allowed?

**Comment by clagms** [comment]
I see what you mean. For debugging purposes it might be useful, so ok. Thanks for improving this! Pls don't forget to push the changes :)

**Comment by masoud-najafi** [comment]
With a Fixed step solver you get this.
image
But for a variable-step solver you need more robustness in the model to make it work correctly..

**Comment by clagms** [comment]
- [ ] It would make sense to add a Sensor and Actuation FMUs, with shared clock connections.

**Comment by masoud-najafi** [comment]
> * [ ]  It would make sense to add a Sensor and Actuation FMUs, with shared clock connections.
@clagms 
Do you mean a sensor with a LPF followed by a clocked Sample/Hold and an actuator followed by a zero-order hold? good idea!

BTW, did you look at the proposed changes for the  supervisor.fmu?

**Comment by clagms** [comment]
> * [x]  It would make sense to add a Sensor and Actuation FMUs, with shared clock connections.

Cancelling this for this PR. Can be made in a future PR.

**Comment by clagms** [comment]
@masoud-najafi I sent you an email for us to catch up on the above changes, and especially discuss the changes to Supervisor.

**Comment by masoud-najafi** [comment]
> @masoud-najafi I sent you an email for us to catch up on the above changes, and especially discuss the changes to Supervisor.

Hi Claudio,
OK. Once finished, please send me the latest version of FMUs to check them on my side too.
Best,
Masoud

**Comment by clagms** [comment]
@t-sommer , @KarlWernersson , @masoud-najafi , I'll contact you via email to try and arrange a session to go over the code, take note of what needs to be improved, and then push this in.

I think we are very close, with the latest push being helped by @masoud-najafi .

**Comment by clagms** [comment]
We are implementing support for FMI 3.0 in the Maestro tool, but we're not yet at the stage where we can test these FMUs. 

That is however independent of this PR. I need a code review with others before merging this.

I'll join the next FMI Design meeting to discuss this. Thanks for following up!

**Comment by t-sommer** [comment]
I've set up the scaffolding for the Synchronous Clock FMUs in https://github.com/clagms/synchronous-clock-fmus. Please move the code there, so we can close this PR.

**Comment by t-sommer** [comment]
Development continues on https://github.com/clagms/synchronous-clock-fmus.

**Pull request closed without merging**